server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -252,28 +258,79 @@ private void loadSnapshotInfos(SnapshotsInProgress snapshotsInProgress, String r
                 snapshotSet.add(new SnapshotInfo(entry));
             }
         }
-        // then, look in the repository
+        // then, look in the repository if there's any matching snapshots left
+        final List<SnapshotInfo> snapshotInfos;
+        if (snapshotIdsToIterate.isEmpty()) {
+            snapshotInfos = Collections.emptyList();
+        } else {
+            snapshotInfos = Collections.synchronizedList(new ArrayList<>());
+        }
+        final ActionListener<Collection<Void>> allDoneListener = listener.delegateFailure((l, v) -> {
+            final ArrayList<SnapshotInfo> snapshotList = new ArrayList<>(snapshotInfos);
+            snapshotList.addAll(snapshotSet);
+            CollectionUtil.timSort(snapshotList);
+            listener.onResponse(unmodifiableList(snapshotList));
+        });
+        if (snapshotIdsToIterate.isEmpty()) {
+            allDoneListener.onResponse(Collections.emptyList());
+            return;
+        }
+        // put snapshot info downloads into a task queue instead of pushing them all into the queue to not completely monopolize the
+        // snapshot meta pool for a single request
+        final int workers = Math.min(threadPool.info(ThreadPool.Names.SNAPSHOT_META).getMax(), snapshotIdsToIterate.size());
+        final Executor executor = threadPool.executor(ThreadPool.Names.SNAPSHOT_META);
+        final BlockingQueue<SnapshotId> queue = new LinkedBlockingQueue<>(snapshotIdsToIterate);
+        final ActionListener<Void> workerDoneListener = new GroupedActionListener<>(allDoneListener, workers).delegateResponse((l, e) -> {
+            queue.clear(); // Stop fetching the remaining snapshots once we've failed fetching one since the response is an error response
+                           // anyway in this case
+            l.onFailure(e);
+        });
         final Repository repository = repositoriesService.repository(repositoryName);
-        for (SnapshotId snapshotId : snapshotIdsToIterate) {
+        for (int i = 0; i < workers; i++) {
+            getOneSnapshotInfo(
+                    ignoreUnavailable,
+                    repository,
+                    queue,
+                    snapshotInfos,
+                    task,
+                    executor,
+                    workerDoneListener
+            );
+        }
+    }
+
+    private void getOneSnapshotInfo(boolean ignoreUnavailable,
+                                    Repository repository,
```

Comment:
We could I guess but it's not going to be a big cleanup/win since this situation is somewhat broken to begin with. 
The behavior of the repository after close isn't well defined currently. Depending on the repo implementation the requests can either start failing or in case of `FsRepository` will just keep going actually because close is a noop there.
Might be worth just fixing that in general at some point?

Reply:
I agree but I still think it is worth not mixing thrown exceptions and listeners here.

Reply:
oh :facepalm: now I get your comment. Sorry, I completely misread it for no good reason :( => Fix coming right up.

Reply:
I pushed fb55daa9562 to address this now :) I went with this instead of looking up the repo in the loop, because the latter would be caught and suppressed by `ignoreUnvailable` which I found weird (albeit practically irrelevant).

Reply:
fb55daa looks good, thanks! And sorry if I wasn't clear at first :)